### PR TITLE
feat: watch resource with selectors

### LINKED ISSUE #1622: watch resource with selector
*Description*:
in order to optimize EG occupy high memory,  #1610 supports EG watching the spec namespace. but the users need to list all namespaces that need watching.  It's not convenient, we can watch the namespace that has some label.  IMO should watch the namespace with the same label, like `shared-gateway-access`  https://gateway-api.sigs.k8s.io/guides/multiple-ns/

xref: https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.15.0/pkg/cache#Options

cc @arkodg

### PR DESCRIPTION
**Pull request by den3tsou**
**What this PR does / why we need it**:
This PR is to provides a handy way for users to watch namespaces with certain labels. Envoy Gateway currently only supports specifying namespaces to watch which is not convenient for users.

Add namespaceSelectors watch mode for Kubernetes provider. Users will need to specify `EnvoyGateway.Provider.Type` and precisely one of `EnvoyGateway.Provider.Kubernetes.Wach.Namespaces` and `EnvoyGateway.Provider.Kubernetes.Wach.NamespaceSelectors` to set the KuberNtes wach mode.

The namespaceSelectors doesn't change the namespace informers watch. The informer still watches all namespaces. The events which have Objects that are not under namesapces with the labels set by `NamespaceSelectors` are filtered out.

**Which issue(s) this PR fixes**:
Fixes #1622

### CODE REVIEW

File: internal/provider/kubernetes/kubernetes.go
Comment (den3tsou) [inline internal/provider/kubernetes/kubernetes.go:45]: There doesn't seem to be any validation in place. Do I need to add it in this PR?

File: internal/provider/kubernetes/predicates.go
Comment (den3tsou) [inline internal/provider/kubernetes/predicates.go:59]: Could anyone help me to understand the common practices in Enovy Gateway to handle errors in predicate?

File: internal/provider/kubernetes/predicates_test.go
Comment (den3tsou) [inline internal/provider/kubernetes/predicates_test.go:67]: Is this the only place to add tests? Do I need to add tests anywhere else?

File: api/config/v1alpha1/envoygateway_types.go
Comment (arkodg) [inline api/config/v1alpha1/envoygateway_types.go:171]:
```suggestion
	KubernetesWatchModeTypeNamespaceSelectors = "NamespaceSelectors"
```

File: api/config/v1alpha1/envoygateway_types.go
Comment (arkodg) [inline api/config/v1alpha1/envoygateway_types.go:167]:
```suggestion
	KubernetesWatchModeTypeNamespaces = "Namespaces"
```

File: internal/provider/kubernetes/predicates.go
Comment (arkodg) [reply internal/provider/kubernetes/predicates.go:59]: logging an error and returning `false` is fine

File: internal/provider/kubernetes/predicates_test.go
Comment (arkodg) [reply internal/provider/kubernetes/predicates_test.go:67]: a good end to end test to add would be to implement this example https://gateway-api.sigs.k8s.io/guides/multiple-ns/ in our e2e https://github.com/envoyproxy/gateway/tree/main/test/e2e/tests

Comment (den3tsou) [reply internal/provider/kubernetes/predicates_test.go:67]: Hi @arkodg, would you please help me understand how to test the predicate behaviour with e2e tests?  Is there any API I could hit in order to know that certain events are filtered out?

Comment (arkodg) [reply internal/provider/kubernetes/predicates_test.go:67]: based on https://gateway-api.sigs.k8s.io/guides/multiple-ns/#shared-gateway you could ensure e2e routing works when `default` ns has label `shared-gateway-access: "true"` set and fails when it is not ns

Comment (den3tsou) [reply internal/provider/kubernetes/predicates_test.go:67]: Hi @arkodg, this is doable but difficult. There is no easy way to reconfig Envoy Gateway in the current testing structure inlcuding 1) apply new ConfigMap 2) restart Envoy Gateway which is needed for this test. Adding those changes will make this PR too big. What's your thoughts on this?

Comment (arkodg) [reply internal/provider/kubernetes/predicates_test.go:67]: okay, then can we add some tests in https://github.com/envoyproxy/gateway/blob/main/internal/provider/kubernetes/kubernetes_test.go that highlight that when we are in this special mode only specific resources are reconciled ?

Comment (den3tsou) [reply internal/provider/kubernetes/predicates_test.go:67]: Hey @arkodg , do you mind elaborating more on what you meant? I've gone through the source code to see if there are any interfaces I could use in order to test the behaviour. Unfortunately, the config go straight into controller.controller.Watch which is implementation details which is not possible to test. Do I miss anything here?

Comment (arkodg) [reply internal/provider/kubernetes/predicates_test.go:67]: hope this test helps https://github.com/envoyproxy/gateway/blob/07297df653e96751b5d1adffdc06c00b92aa431a/internal/provider/kubernetes/kubernetes_test.go#L1669

Comment (den3tsou) [reply internal/provider/kubernetes/predicates_test.go:67]: This test is a bit different from what I am doing because the other watch mode changes the client, so using the client to query can test the behavior. But the one I am currently working on doesn't change the behavior of the client. It changes the behavior of informers. I guess I miss something here. I will have a look again. If this doesn't work, I think the only solution is to change the test structure of E2E tests so that updating Envoy Gateway is possible

Comment (arkodg) [reply internal/provider/kubernetes/predicates_test.go:67]: for the verification steps, instead of using the client you can peek into the `ProviderResources` message to see if the resource should to shouldnt exist

Comment (den3tsou) [reply internal/provider/kubernetes/predicates_test.go:67]: After the investigation, I realise that adding predicate will not help in this case because reconciling will try to obtain everything for whatever event comes in. The direction of this PR is not correct. I will dig into more to see what I need to do in order to implement this feature within reconciling.

Comment (den3tsou) [reply internal/provider/kubernetes/predicates_test.go:67]: Hi @arkodg , I would like to get your opinion before proceeding. After the investigation, addtional logic is needed to be added into functions including `processGateways`, `processTLSRoutes`, `processHTTPRoutes`, `processGRPCRoutes`, `processTCPRoutes` and `processUDPRoutes` to have additional check to make sure that only resources in the namespaces with certain labels are processed. I can see that this will increase the complexity of the codebase. I just wonder if you are happy with me going head? or you have other solutions in mind.

Comment (arkodg) [reply internal/provider/kubernetes/predicates_test.go:67]: can we do it in a way that doesnt add too much complexity ? but the goal is what you described - `make sure that only resources in the namespaces with certain labels are processed`

Comment (den3tsou) [reply internal/provider/kubernetes/predicates_test.go:67]: This is a bit tricky. I will share what I've found, so you could help me to brainstorm some ideas.

There are two approach available at this moment

-  Initialise the k8s client to obtain all namespaces with certain labels
    * Pros:
       * No complex change for reconciler because configuring Manager is enough
    * Cons:
       * Users need to restart Envoy Gateway controller every time they add update the namespaces unless there is another goroutine or anything similar to check the namespace regularly and update the Manger config accordingly
- Use predicate and update all of logic inside reconciler to check if the objects' namespaces have certain labels
    * Pros:
       * Users don't need to redeploy Envoy Gateway controller every time they update namespaces
    * Cons:
       * All logic inside reconciler uses manager.client needs to have additional logic to check if the objects are under namepsace with certain labels. This may cause performance issue because every object will need to be checked.

These two solutions are what I've found so far. I've checked the `controller-runtime` to see if it has any features that I could workaround. Unfortunately, I coudln't find any solutions yet. Do you have any ideas?

Comment (den3tsou) [reply internal/provider/kubernetes/predicates_test.go:67]: Hi @arkodg  This is a bit tricky. I will share what I've found, so you could help me to brainstorm some ideas.

There are two approach available at this moment

-  Initialise the k8s client to obtain all namespaces with certain labels
    * Pros:
       * No complex change for reconciler because configuring Manager is enough
    * Cons:
       * Users need to restart Envoy Gateway controller every time they add update the namespaces unless there is another goroutine or anything similar to check the namespace regularly and update the Manger config accordingly
- Use predicate and update all of logic inside reconciler to check if the objects' namespaces have certain labels
    * Pros:
       * Users don't need to redeploy Envoy Gateway controller every time they update namespaces
    * Cons:
       * All logic inside reconciler uses manager.client needs to have additional logic to check if the objects are under namepsace with certain labels. This may cause performance issue because every object will need to be checked.

These two solutions are what I've found so far. I've checked the `controller-runtime` to see if it has any features that I could workaround. Unfortunately, I coudln't find any solutions yet. Do you have any ideas?

Comment (den3tsou) [reply internal/provider/kubernetes/predicates_test.go:67]: Hi @arkodg , any update on this. I am fine to decline this PR and work on other tasks. I've spent fair bit of time finding other solution. But this is more related to how reconciler works, there is no work around for it. All of code related to pull data from api server will need to be updated.

Comment (den3tsou) [reply internal/provider/kubernetes/predicates_test.go:67]: Thanks 👍 . I will start to implement it :-D

File: api/config/v1alpha1/envoygateway_types.go
Comment (arkodg) [inline api/config/v1alpha1/envoygateway_types.go:197]: needs a default tag here similar to https://github.com/envoyproxy/gateway/blob/91c0888ad63631440286264f18f1bd8d2ccab4cd/api/config/v1alpha1/envoyproxy_types.go#L169

File: internal/provider/kubernetes/routes.go
Comment (arkodg) [inline internal/provider/kubernetes/routes.go:589]: can all this code be consolidated by using a helper func with a client.Object signature ?

File: internal/provider/kubernetes/predicates.go
Comment (arkodg) [inline internal/provider/kubernetes/predicates.go:53]:
```suggestion
				err, "failed to get Namespace labels",
```

File: internal/provider/kubernetes/controller.go
Comment (arkodg) [inline internal/provider/kubernetes/controller.go:1233]: why is `ls` needed here, can't you just use `r.namespaceLabels` as a way to check

File: api/config/v1alpha1/envoygateway_types.go
Comment (den3tsou) [reply api/config/v1alpha1/envoygateway_types.go:197]: Do I miss anything here? This is an optional field. It has to be defined with either `Namespaces` or `NamespaceSelectors`

File: internal/provider/kubernetes/routes.go
Comment (den3tsou) [reply internal/provider/kubernetes/routes.go:589]: ahhh. Good catch. I didn't realise that `client.Object` can be used!

File: internal/provider/kubernetes/controller.go
Comment (den3tsou) [reply internal/provider/kubernetes/controller.go:1233]: good catch. updated

File: internal/provider/kubernetes/predicates.go
Comment (den3tsou) [reply internal/provider/kubernetes/predicates.go:53]: Fixed

File: internal/provider/kubernetes/routes.go
Comment (den3tsou) [reply internal/provider/kubernetes/routes.go:589]: Unfortunately, `AuthenticaionFilter`, RateLimitFilter` and `unstructured.Unstructured` don't implment the common interface. The alternative I could do is to add a NamespaceGetter interface to simplify the code.

Comment (arkodg) [reply internal/provider/kubernetes/routes.go:582]: you could use existing code for them, and add a `TODO`

File: api/config/v1alpha1/envoygateway_types.go
Comment (den3tsou) [reply api/config/v1alpha1/envoygateway_types.go:197]: Hi @arkodg, do you mind helping me understand this?

**Comment by den3tsou** [comment]
Hi @arkodg , I finally complete the integration tests. Would you please have a look again?

File: api/config/v1alpha1/envoygateway_types.go
Comment (arkodg) [reply api/config/v1alpha1/envoygateway_types.go:197]: this should be `// +kubebuilder:default=Namespaces`

**Review by arkodg — APPROVED** [review]
one nit, else looks good, thanks @den3tsou !

**Comment by arkodg** [comment]
lets merge this one after https://github.com/envoyproxy/gateway/pull/1534 has merged to reduce conflicts

**Review by AliceProxy — APPROVED** [review]
lgtm. @arkodg I see no conflicts after merging #1534 so I'm gonna go ahead and merge this now to prevent additional churn on this PR. Any outstanding nits can be resolved via a follow-up PR.

**Pull request merged**
